vm/systemSmartContracts/validator.go

```diff
@@ -1069,46 +1075,73 @@ func (v *validatorSC) stake(args *vmcommon.ContractCallInput) vmcommon.ReturnCod
 		}
 	}
 
-	if !v.isNumberOfNodesTooHigh(registrationData) {
-		v.activateStakingFor(
-			blsKeys,
-			registrationData,
-			validatorConfig.NodePrice,
-			registrationData.RewardAddress,
-			args.CallerAddr,
-		)
-	} else {
-		numRegisteredBlsKeys := int64(len(registrationData.BlsPubKeys))
+	v.activateNewBLSKeys(registrationData, blsKeys, newKeys, &validatorConfig, args)
+
+	err = v.saveRegistrationData(args.CallerAddr, registrationData)
+	if err != nil {
+		v.eei.AddReturnMessage("cannot save registration data: error " + err.Error())
+		return vmcommon.UserError
+	}
+
+	return vmcommon.Ok
+}
+
+func (v *validatorSC) activateNewBLSKeys(
+	registrationData *ValidatorDataV2,
+	blsKeys [][]byte,
+	newKeys [][]byte,
+	validatorConfig *ValidatorConfig,
+	args *vmcommon.ContractCallInput,
+) {
+	numRegisteredBlsKeys := len(registrationData.BlsPubKeys)
+	numNodesTooHigh := v.activateStakingFor(
+		blsKeys,
+		newKeys,
+		registrationData,
+		validatorConfig.NodePrice,
+		registrationData.RewardAddress,
+		args.CallerAddr,
+	)
+
+	if numNodesTooHigh && len(blsKeys) > 0 {
```

Comment:
Can it ever happen that len(blsKeys) is 0 here?

Reply:
yes. in 99% of cases. blsKeys are there when automatic activation is on for delegation contract. Or the validator explicitely wants to stake more nodes.